#26: Added user guide (#67)
exasol/advanced-analytics-framework: doc/user_guide/user_guide.md

# Advanced Analytics Framework User Guide

The Advanced Analytics Framework (AAF) provides a general framework to implement complex data analysis algorithms with Exasol. Users can use the features of AAF in their custom implementations.

## Table of Contents

* [Getting Started](#getting-started)
* [Setup](#setup)
* [Usage](#usage)
* [Implementation of Algorithms](#implementation-of-algorithms)

## Getting Started

* Exasol database
  * The Exasol cluster must already be running with version 7.1 or later.
  * Database connection information and credentials are needed.
* TODO: Connection
* TODO: Algorithm implementation

## Setup

### The Python Package

#### Download The Python Wheel Package

The latest version of the python package of the framework can be downloaded from the Releases in GitHub Repository (see [the latest release](https://github.com/exasol/advanced-analytics-framework/releases/latest)).

Please download the archive `advanced_analytics_framework.whl`.

#### Install The Python Wheel Package

The following command installs the package `advanced-analytics-framework` from [pypi](https://pypi.org):

```bash
pip install exasol-advanced-analytics-framework
```

> ### The Pre-built Script Language Container (SLC)

Reviewer: I think, we need to try something here, if you look at _Implementation of Algorithms_ , you see a user can configure a UDF, maybe the class can be implemented in the UDF, then a prebuilt container would still make sense.

Reviewer: But, maybe it still makes sense, that this framework could be used as a dependency for a separate extension.

Reviewer 2: I think we first need to decide whether AAF should offer a pre-built SLC or not.

If yes, then we probably need to add this pre-built SLC or some logic in order to build it. 
This will probably be addressed in a separate ticket and hence out of scope for the current ticket. 

In any case, I propose to either remove or update the current section in the user guide until the decision and the potentially required changes are applied.

Reviewer: if we offer an prebuilt SLC depends on, if my idea works or not. If it works, it makes sense, otherwise not. So, your task is probably trying it out. Built a custom UDF which calls the UDFQueryHandlerRunner with a custom UDFQueryHander.

Reviewer 2: Current experiments showed 
* UDF is in Python module `builtin`
* SLC could be uploaded to BucketFS of database in Docker Container

# User Guide

The goal of this library is to provide a general framework to implement complex 
data analysis algorithms with Exasol. This framework provides certain features 
to users in which they are able to run their implementations.

## Table of Contents

- [Getting Started](#getting-started)
- [Setup](#setup)
- [Usage](#usage)
- [Implementation of Algorithms](#implementation-of-algorithms)

## Getting Started
- Exasol DB
  - The Exasol cluster must already be running with version 7.1 or later.
  - DB connection information and credentials are needed.

> - TODO: Connection

Reviewer: Example for a bucketfs connection via create connection.

Reviewer 2: See next push.

> - TODO: Algorithm implementation

Reviewer: obsolete

Reviewer 2: Regarding the Algorithm implementation, I had the same impression.
For the connection: see next push.

# Advanced Analytics Framework User Guide

The Advanced Analytics Framework (AAF) enables implementing complex data analysis algorithms with Exasol. Users can use the features of AAF in their custom implementations.

## Table of Contents

* [Setup](#setup)
* [Usage](#usage)
* [Custom Algorithms](#custom-algorithms)

## Setup

### Exasol database

* The Exasol cluster must already be running with version 7.1 or later.
* Database connection information and credentials are needed for the database itself as also for the BucketFS.

### BucketFS Connection

AAF employs some Lua scripts and User Defined Functions (UDFs).  The Lua scripts are orchestrating the UDFs while the UDFs are performing the actual analytic functions.

> AAF keeps a common state of execution and passes input data and results between Lua and UDFs via files in the Bucket File System (BucketFS) of the Exasol database.

Reviewer: This is basically a minimalistic explanation, maybe it should go into its own section

Reviewer 2: See next push.

| `onprem` | `<BUCKET_NAME>`      | yes       | `default`      | Name of the Bucket                                                 |
| `onprem` | `<PATH>`             | -         | (empty / root) | Path inside the Bucket                                             |
| `onprem` | `<VERIFY>`           | -         | `true`         | Whether to apply TLS security to the connection                    |
| `saas`   | `<SAAS_ACCOUNT_ID>`  | yes       | -              | Account ID for accessing an SaaS database instance                 |
| `saas`   | `<SAAS_DATABASE_ID>` | yes       | -              | Database ID of an Exasol SaaS database instance                    |
| `saas`   | `<SAAS_PAT>`         | yes       | -              | Personal access token for accessing an SaaS database instance      |

### AAF Python Package

The latest version of AAF can be obtained from [pypi](https://pypi.org), see also the [releases on GitHub](https://github.com/exasol/advanced-analytics-framework/releases):

```bash
pip install exasol-advanced-analytics-framework
```

### Script Language Container (SLC)

> Exasol executes User Defined Functions (UDFs) in a isolated Container whose root filesystem is derived Script Language Container (SLC).

Reviewer: ```suggestion
Exasol executes User Defined Functions (UDFs) in a isolated Container whose root filesystem is derived from a Script Language Container (SLC).
```

> Running the AAF requires a custom SLC. The following command

Reviewer: ```suggestion
Running the AAF requires a SLC. The following command
```

custom is here not correct, because it is not always customized, it could be also prebuilt.

|------------------------------|-----------|-------------------------------------------------------------------------------|
| `<CLASS_NAME>`               | yes       | Name of the query handler class                                               |
| `<CLASS_MODULE>`             | yes       | Module name of the query handler class                                        |
| `<CLASS_PARAMETERS>`         | yes       | Parameters of the query handler class encoded as string                       |
| `<UDF_NAME>`                 | -         | Name of Python UDF script that contains the algorithm implemented by the user |
| `<UDF_DB_SCHEMA>`            | -         | Schema name where the UDF script is deployed                                  |
| `<BUCKETFS_CONNECTION_NAME>` | yes       | BucketFS connection name which is used to create temporary bucketfs files     |
| `<BUCKETFS_DIRECTORY>`       | yes       | Directory in BucketFS for the temporary bucketfs files                        |
| `<TEMP_DB_SCHEMA>`           | yes       | Database Schema for temporary database objects, e.g. tables                   |

> Please take care to provide a string value for `<CLASS_PARAMETERS>`.  Simple data types like `float`, `int`, `bool` will be converted to a String, while a Json object or an array is represented as a string value with an unusable reference, e.g.  `table: 0x14823bd38580`.

Reviewer: We should that the json object problem will be fixed

Reviewer 2: See next push.

## Custom Algorithms

### Deployment Options

Using the AAF requires to implement a custom algorithm using one of the following alternatives
* Adhoc implementation within a UDF
* Build a custom extension

#### Building a custom extension

> * Create a python package that depends on the python package of the AAF and that implements the query handler.of the custom algorithm and its factory class.

Reviewer: ```suggestion
* Create a python package that depends on the python package of the AAF and that implements the query handler of the custom algorithm and its factory class.
```

* Create an associated SLC which has the python package installed
  * GitHub repository [python-extension-common](https://github.com/exasol/python-extension-common/) provides more detailed documentation and automation.
* Leave out entry `udf` from the json input to use the default UDF.
* Values `<CLASS_MODULE>` and `<CLASS_NAME>` must reflect the _module_ and _class name_ of the `QueryHandler` implemented in the custom SLC.

> ### Implementing a Custom Algorithm as Example Query Handler

Reviewer: ```suggestion
### Example Implementation of a Custom Algorithm as Query Handler
```

Each algorithm should extend the `UDFQueryHandler` abstract class and then implement the following methods:
* `start()`: This method is called at the first execution of the query hander, that is, in the first iteration. It returns a result object: Either _Finish_ or _Continue_.
  * The _Finish_ result object contains the final result of the implemented algorithm.
  * The _Continue_ object contains the query list that will be executed before the next iteration and whose results are used as input for the next iteration.
* `handle_query_result()`: This method is called at the following iterations to handle the result of the queries of the previous iteration.

> Here is an example class definition using an adhoc implementation within the UDF. The example uses the module `builtins` and dynamically adds ExampleQueryHandler and ExampleQueryHandlerFactory to it.

Reviewer: Add section Example

import builtins
builtins.ExampleQueryHandler=ExampleQueryHandler # required for pickle

class ExampleQueryHandlerFactory:
      def create(self, parameter: str, query_handler_context: QueryHandlerContext):
          return builtins.ExampleQueryHandler(parameter, query_handler_context)

builtins.ExampleQueryHandlerFactory=ExampleQueryHandlerFactory

from exasol_advanced_analytics_framework.udf_framework.query_handler_runner_udf \
    import QueryHandlerRunnerUDF

udf = QueryHandlerRunnerUDF(exa)

def run(ctx):
    return udf.run(ctx)
/

EXECUTE SCRIPT MY_SCHEMA.AAF_RUN_QUERY_HANDLER('{
    "query_handler": {
        "factory_class": {
            "module": "builtins",
            "name": "ExampleQueryHandlerFactory"
        },
        "parameter": "bla-bla",
        "udf": {
            "schema": "MY_SCHEMA",
            "name": "MY_QUERY_HANDLER_UDF"
        }
    },
    "temporary_output": {
        "bucketfs_location": {
            "connection_name": "BFS_CON",
            "directory": "temp"
        },
        "schema_name": "TEMP_SCHEMA"
    }
}');
```

The figure below illustrates the execution of this algorithm implemented in class `ExampleQueryHandler`.

> * When method `start()` is called, it executes two queries and an additional `input_query` to obtain the next state.

Reviewer: ```suggestion
* When method `start()` is called, it executes two queries and an additional `input_query` to obtain the input for the next .iteration.
```

* After the first iteration is completed, the framework calls method the `handle_query_result` with the `query_result` of the `input_query` of the previous iteration.

> In this example, the algorithm is finished at this state and returns 2<sup>_return value_</sup> as final result.

Reviewer: ```suggestion
In this example, the algorithm is finished at this iteration and returns 2<sup>_return value_</sup> as final result.
```

| `onprem` | `<BUCKET_NAME>`      | yes       | `default`      | Name of the Bucket                                                 |
| `onprem` | `<PATH>`             | -         | (empty / root) | Path inside the Bucket                                             |
| `onprem` | `<VERIFY>`           | -         | `true`         | Whether to apply TLS security to the connection                    |
| `saas`   | `<SAAS_ACCOUNT_ID>`  | yes       | -              | Account ID for accessing an SaaS database instance                 |
| `saas`   | `<SAAS_DATABASE_ID>` | yes       | -              | Database ID of an Exasol SaaS database instance                    |
| `saas`   | `<SAAS_PAT>`         | yes       | -              | Personal access token for accessing an SaaS database instance      |

### AAF Python Package

The latest version of AAF can be obtained from [pypi](https://pypi.org), see also the [releases on GitHub](https://github.com/exasol/advanced-analytics-framework/releases):

```bash
pip install exasol-advanced-analytics-framework
```

### Script Language Container (SLC)

> Exasol executes User Defined Functions (UDFs) in an isolated Container called _SLC_.

Reviewer: ```suggestion
Exasol executes User Defined Functions (UDFs) in a isolated Container whose root filesystem is derived Script Language Container (SLC).
```

Maybe add a link to SLC repo

Running the AAF requires a custom SLC. The following command
* downloads the specified version `<VERSION>` (preferrably the latest) of a prebuilt AAF SLC from the [AAF releases](https://github.com/exasol/advanced-analytics-framework/releases/latest) on GitHub,
* uploads the file into the BucketFS,
* and registers it to the database.

```shell
LANGUAGE_ALIAS=PYTHON3_AAF
python -m exasol_advanced_analytics_framework.deploy language-container \
    --dsn <DB_HOST:DB_PORT> \
    --db-user <DB_USER> \
    --db-pass <DB_PASSWORD> \
    --bucketfs-name <BUCKETFS_NAME> \
    --bucketfs-host <BUCKETFS_HOST> \
    --bucketfs-port <BUCKETFS_PORT> \
    --bucketfs-user <BUCKETFS_USER> \
    --bucketfs-password <BUCKETFS_PASSWORD> \
    --bucket <BUCKETFS_NAME> \
    --path-in-bucket <PATH_IN_BUCKET> \
    --version <VERSION> \

>     --language-alias "$LANGUAGE_ALIAS"

Reviewer: Hm, using $ and <> looks weird, you probably want to show that the language alias is needed in two places

Reviewer 2: Yes, I already added a back-reference in the next example commandline, But see next push for
* consistently only use `$` references
* add a short rationale why environment variable `LANGUAGE_ALIAS` is set before

| `<UDF_NAME>`                 | -         | Name of Python UDF script that contains the algorithm implemented by the user |
| `<UDF_DB_SCHEMA>`            | -         | Schema name where the UDF script is deployed                                  |
| `<BUCKETFS_CONNECTION_NAME>` | yes       | BucketFS connection name which is used to create temporary bucketfs files     |
| `<BUCKETFS_DIRECTORY>`       | yes       | Directory in BucketFS for the temporary bucketfs files                        |
| `<TEMP_DB_SCHEMA>`           | yes       | Database Schema for temporary database objects, e.g. tables                   |

Please take care to provide a string value for `<CLASS_PARAMETERS>`.  Simple data types like `float`, `int`, `bool` will be converted to a String, while a Json object or an array is represented as a string value with an unusable reference, e.g.  `table: 0x14823bd38580`.

## Custom Algorithms

### Deployment Options

Using the AAF requires to implement a custom algorithm using one of the following alternatives
* Adhoc implementation within a UDF
* Build a custom extension

#### Building a custom extension

> * Create a python package implementing the query handler.

Reviewer: ```suggestion
* Create a python package that depends on the python package of the AAF and that implements the query handler.of the custom algorithm.
```

* Create an associated SLC which has the python package installed
  * GitHub repository [python-extension-common](https://github.com/exasol/python-extension-common/) provides more detailed documentation and automation.

> * Leave out entry `udf` from the json input to use the default UDF. <!--

Reviewer: Your version is better, you can remove the original comment

Reviewer 2: Thanks :pray:

Original comment:
* Leave out entry `udf` `udf_name` and `udf_schema` from the json input to use the default UDF
-->
* Values `<CLASS_MODULE>` and `<CLASS_NAME>` must reflect the _module_ and _class name_ of the `QueryHandler` implemented in the custom SLC.

### Implementation of the Custom Algorithm

Each algorithm should extend the `UDFQueryHandler` abstract class and then implement the following methods:
* `start()`: This method is called at the first execution of the query hander, that is, in the first iteration. It returns a result object: Either _Finish_ or _Continue_.
  * The _Finish_ result object contains the final result of the implemented algorithm.

>   * The _Continue_ object contains the query list that will be executed before the next iteration and whose results are used as input fo the next iteraton.

Reviewer: ```suggestion
  * The _Continue_ object contains the query list that will be executed before the next iteration and whose results are used as input for the next iteraton.
```

      def create(self, parameter: str, query_handler_context: QueryHandlerContext):
          return builtins.ExampleQueryHandler(parameter, query_handler_context)

builtins.ExampleQueryHandlerFactory=ExampleQueryHandlerFactory

from exasol_advanced_analytics_framework.udf_framework.query_handler_runner_udf \
    import QueryHandlerRunnerUDF

udf = QueryHandlerRunnerUDF(exa)

def run(ctx):
    return udf.run(ctx)
/

EXECUTE SCRIPT MY_SCHEMA.AAF_RUN_QUERY_HANDLER('{
    "query_handler": {
        "factory_class": {
            "module": "builtins",
            "name": "ExampleQueryHandlerFactory"
        },
        "parameter": "bla-bla",
        "udf": {
            "schema": "MY_SCHEMA",
            "name": "MY_QUERY_HANDLER_UDF"
        }
    },
    "temporary_output": {
        "bucketfs_location": {
            "connection_name": "BFS_CON",
            "directory": "temp"
        },
        "schema_name": "TEMP_SCHEMA"
    }
}');
```

The figure below illustrates the execution of this algorithm implemented in class `ExampleQueryHandler`.
* When method `start()` is called, it executes two queries and an additional `input_query` to obtain the next state.

> * After the first iteration is completed, the framework calls method `handle_query_result` with the `query_result` of the `input_query` of the previous iteration.

Reviewer: ```suggestion
* After the first iteration is completed, the framework calls method the `handle_query_result` with the `query_result` of the `input_query` of the previous iteration.
```

> In this example, the algorithm is finished at this state and returns 2<sup>_return value_</sup> as final result.

Reviewer: It actually returns

Finish(result=f"Assertion of the final result: 32 == {result}

Reviewer: Code and image are inconsistent, please adapt the code to the image for the moment

Reviewer 2: See next push

| `onprem` | `<PATH>`             | -         | (empty / root) | Path inside the Bucket                                             |
| `onprem` | `<VERIFY>`           | -         | `true`         | Whether to apply TLS security to the connection                    |
| `saas`   | `<SAAS_ACCOUNT_ID>`  | yes       | -              | Account ID for accessing an SaaS database instance                 |
| `saas`   | `<SAAS_DATABASE_ID>` | yes       | -              | Database ID of an Exasol SaaS database instance                    |
| `saas`   | `<SAAS_PAT>`         | yes       | -              | Personal access token for accessing an SaaS database instance      |

### AAF Python Package

The latest version of AAF can be obtained from [pypi](https://pypi.org), see also the [releases on GitHub](https://github.com/exasol/advanced-analytics-framework/releases).

The following command installs the AAF from pypi:

```bash
pip install exasol-advanced-analytics-framework
```

### Script Language Containers (SLCs)

Exasol executes User Defined Functions (UDFs) in an isolated Script Language Container (SLCs).
Running the AAF requires a custom SLC.

> #### Build the SLC

Reviewer: I think, this section we can move to the developer guide, because we will release the prebuilt slc

Reviewer 2: See next push

2. Install poetry
3. Use the AAF developer commands for building and deploying custom SLCs into Exasol.

```shell
poetry run nox build_language_container
```

See [Install the SLC](#install-the-slc).

#### Download a Pre-built SLC

As an alternative to building SLC yourself you can also download a prebuilt AAF SLC from the [AAF releases](https://github.com/exasol/advanced-analytics-framework/releases/latest) on GitHub.

#### Install the SLC

Installing the SLC requires loading the container file into the BucketFS and registering it to the database:

```shell
SLC_FILE=.slc/exasol_advanced_analytics_framework_container_release.tar.gz
LANGUAGE_ALIAS=PYTHON3_AAF
python -m exasol_advanced_analytics_framework.deploy language-container \
    --dsn <DB_HOST:DB_PORT> \
    --db-user <DB_USER> \
    --db-pass <DB_PASSWORD> \
    --bucketfs-name <BUCKETFS_NAME> \
    --bucketfs-host <BUCKETFS_HOST> \
    --bucketfs-port <BUCKETFS_PORT> \
    --bucketfs-user <BUCKETFS_USER> \
    --bucketfs-password <BUCKETFS_PASSWORD> \
    --bucket <BUCKETFS_NAME> \
    --path-in-bucket <PATH_IN_BUCKET> \
    --language-alias "$LANGUAGE_ALIAS" \

>     --container-file "$SLC_FILE"

Reviewer: the slc deployer should be able to download the prebuilt slc automatically from github, see https://github.com/exasol/transformers-extension/blob/f31b7ea91db5e560b6207b6efcba52eddfd4c7ce/exasol_transformers_extension/deployment/te_language_container_deployer.py#L11

--container-file is only fallback for special circumstances

Reviewer 2: Ok, I am sorry to maybe have created some confusion.
Yes: AAF deploy CLI has already been ported to the latest PEC version and is able to download the SLC file from the specified GitHub release.

```

### Additional Scripts

Besides the BucketFS connection, the SLC, and the Python package AAF also requires some additional Lua scripts to be created in the Exasol database.

The following command deploys the additional scripts to the specified `DB_SCHEMA` using the `LANGUAGE_ALIAS` of the SLC:

```shell
python -m exasol_advanced_analytics_framework.deploy scripts \
    --dsn <DB_HOST:DB_PORT> \
    --db-user <DB_USER> \
    --db-pass <DB_PASSWORD> \
    --schema <DB_SCHEMA> \
    --language-alias "$LANGUAGE_ALIAS"
```

## Usage

> The entry point of this framework is `AAF_RUN_QUERY_HANDLER` script. This script is simply a query loop which is responsible for executing the implemented algorithm.

Reviewer: We might need to explain that an algorithm is implemented as QueryHandler and what a QueryHandler is.

Reviewer 2: Proposal: Move to separate PR.

> This script takes the necessary parameters to execute the desired algorithm in string json format. The json input includes two main part:

Reviewer: ```suggestion
This script takes the necessary parameters to execute the desired algorithm in string json format. The json input includes two main parts:
```

> * `query_handler` : Details of the algorithm implemented by user.

Reviewer: ```suggestion
* `query_handler` : Details of the algorithm implemented by the user.
```

```sql
EXECUTE SCRIPT AAF_RUN_QUERY_HANDLER('{
    "query_handler": {
        "factory_class": {
            "module": "<CLASS_MODULE>",
            "name": "<CLASS_NAME>"
        },
        "parameters": "<CLASS_PARAMETERS>",
        "udf": {
            "schema": "<UDF_DB_SCHEMA>",
            "name": "<UDF_NAME>"
        }
    },
    "temporary_output": {
        "bucketfs_location": {
            "connection_name": "<BUCKETFS_CONNECTION_NAME>",
            "directory": "<BUCKETFS_DIRECTORY>"
        },
        "schema_name": "<TEMP_DB_SCHEMA>"
    }
}');
```

See [Implementation of Custom Algorithms](#implementation-of-custom-algorithms) for a complete example.

### Parameters

| Parameter                    | Required? | Description                                                                                                    |
|------------------------------|-----------|----------------------------------------------------------------------------------------------------------------|
| `<CLASS_NAME>`               | yes       | Name of the query handler class                                                                                |

> | `<CLASS_MODULE>`             | yes       | Module name of the query handler class <span style="color: red">(should we mention `builtins` here?)</span>    |

Reviewer: the builtins only comes into play when we implement the queryhandler in UDF. We should mention that in the example.

Reviewer 2: See next push.

> | `<CLASS_PARAMETERS>`         | yes       | Parameters of the query handler class encoded as string                                                        |

Reviewer: actually we should check, what happens when this is missing or a json object or array

Reviewer 2: Using an array `"parameters": [ "bla-bla" ],` results in an unusable string containing a reference passed to the UDF: `table: 0x14823bd38580`.

Reviewer 2: I added some explanation to the user guide.

Reviewer 2: See also #196

> | `<UDF_NAME>`                 | -         | Name of Python UDF script including user-implemented algorithm                                                 |

Reviewer: ```suggestion
| `<UDF_NAME>`                 | -         | Name of Python UDF script that contains the algorithm implemented by the user                                                |
```

| `<UDF_DB_SCHEMA>`            | -         | Schema name where the UDF script is deployed                                                                   |

> | `<BUCKETFS_CONNECTION_NAME>` | yes       | BucketFS connection name to create temporary file outputs                                                      |

Reviewer: ```suggestion
| `<BUCKETFS_CONNECTION_NAME>` | yes       | BucketFS connection name to create temporary bucketfs files                                                      |
```

| `<BUCKETFS_DIRECTORY>`       | yes       | Directory in BucketFS for temporary file outputs                                                               |
| `<TEMP_DB_SCHEMA>`           | yes       | Database Schema for temporary database objects, e.g. tables                                                    |

# Implementation of Custom Algorithms

Each algorithm should extend the `UDFQueryHandler` abstract class and then implement the following methods:

> * `start()`: This method is called at the first execution of the framework, that is, in the first iteration. It returns a result object: Either _Finish_ or _Continue_.

Reviewer: ```suggestion
* `start()`: This method is called at the first execution of the query hander, that is, in the first iteration. It returns a result object: Either _Finish_ or _Continue_.
```

  * The _Finish_ result object contains the final result of the implemented algorithm.

>   * The _Continue_ object contains the query list that will be executed for the next state.

Reviewer: ```suggestion
  * The _Continue_ object contains the query list that will be executed before the next iteration and whose results are used as input fo the next iteraton.
```

> * `handle_query_result()`: This method is called at the following iterations to handle the return query.

Reviewer: ```suggestion
* `handle_query_result()`: This method is called at the following iterations to handle the result of the queries of the previous iteration.
```

> 

Reviewer: We probably need to mention usage modes for the AAF,

1) Adhoc implementation of algorithms within a UDF
2) Building an extension
  - means python package which implements the query handler and an associated SLC which has the python package installed 
  - here we only give a link to PEC
  - we give an example how would need to AAF_RUN_QUERY_HANDLER in this case

Reviewer 2: This sentence I don't understand. 

> we give an example how would need to AAF_RUN_QUERY_HANDLER in this case

Reviewer: > This sentence I don't understand.
> 
> > we give an example how would need to AAF_RUN_QUERY_HANDLER in this case

we give an example how would a user need to call AAF_RUN_QUERY_HANDLER in this case

Reviewer 2: Thanks - see next push.

Here is an example class definition:

```python
--/
CREATE OR REPLACE PYTHON3_AAF SET SCRIPT "MY_SCHEMA"."MY_QUERY_HANDLER_UDF"(...)
EMITS (outputs VARCHAR(2000000)) AS

from typing import Union
from exasol_advanced_analytics_framework.udf_framework.udf_query_handler import UDFQueryHandler
from exasol_advanced_analytics_framework.query_handler.context.query_handler_context import QueryHandlerContext
from exasol_advanced_analytics_framework.query_result.query_result import QueryResult
from exasol_advanced_analytics_framework.query_handler.result import Result, Continue, Finish
from exasol_advanced_analytics_framework.query_handler.query.select_query import SelectQuery, SelectQueryWithColumnDefinition
from exasol_data_science_utils_python.schema.column import Column
from exasol_data_science_utils_python.schema.column_name import ColumnName
from exasol_data_science_utils_python.schema.column_type import ColumnType

> -- proposal write parameter into temp table in schema ...

Reviewer: maybe the following scenario

start
1. write parameter+"bucketfs"+time to temp bucketfs file
2. create temp table
3. insert parameter+"table-insert"+time to temp table
4. Return query select from temp table and concat "return"

handle result
1. read result
2. read temp bucketfs file
3. create table from result
4. insert temp bucketfs file content into new table
5. return new table name

Reviewer 2: Proposal: Move to separate PR.

class CustomQueryHandler(UDFQueryHandler):
    def __init__(self, parameter: str, query_handler_context: QueryHandlerContext):
        super().__init__(parameter, query_handler_context)
        self.parameter = parameter
        self.query_handler_context = query_handler_context

    def start(self) -> Union[Continue, Finish[str]]:
        query_list = [
          SelectQuery("SELECT 1 FROM DUAL"),
          SelectQuery("SELECT 2 FROM DUAL")]
        query_handler_return_query = SelectQueryWithColumnDefinition(
            query_string="SELECT 5 AS 'return_column' FROM DUAL",
            output_columns=[
              Column(ColumnName("return_column"), ColumnType("INTEGER"))])

        return Continue(
            query_list=query_list,
            input_query=query_handler_return_query)

    def handle_query_result(self, query_result: QueryResult) -> Union[Continue, Finish[str]]:
        return_value = query_result.return_column
        result = 2 ** return_value
        return Finish(result=f"Assertion of the final result: 32 == {result}")

import builtins

> builtins.CustomQueryHandler=CustomQueryHandler # required for pickle

Reviewer: in a extra PR we need to add a function like register_query_handler_inside_udf and comment that this is only needed when you define the query handler insidde of a UDF

    import QueryHandlerRunnerUDF

udf = QueryHandlerRunnerUDF(exa)

def run(ctx):
    return udf.run(ctx)
/

EXECUTE SCRIPT MY_SCHEMA.AAF_RUN_QUERY_HANDLER('{
    "query_handler": {
        "factory_class": {
            "module": "builtins",
            "name": "CustomQueryHandlerFactory"
        },
        "parameters": "bla-bla",
        "udf": {
            "schema": "MY_SCHEMA",
            "name": "MY_QUERY_HANDLER_UDF"
        }
    },
    "temporary_output": {
        "bucketfs_location": {
            "connection_name": "BFS_CON",
            "directory": "temp"
        },
        "schema_name": "TEMP_SCHEMA"
    }
}');
```

The figure below illustrates the execution of this algorithm implemented in class `ExampleQueryHandler`.
* When method `start()` is called, it executes two queries and an additional `input_query` to obtain the next state.
* After the first iteration is completed, the framework calls method `handle_query_result` with the `query_result` of the `input_query` of the previous iteration.

In this example, the algorithm is finished at this state and returns 2<sup>_return value_</sup> as final result.

> ![Sample Execution](../images/sample_execution.png "Sample Execution")

Reviewer: With the extended scenario, we need to redo the image

Reviewer 2: We will keep the Image for now, please fix typo "Contiue". 
Image will be updated in separate PR, anyway, accordingly to the extended scenario.
